userland/libtock/crt0.c

```diff
@@ -6,23 +6,77 @@ extern int main(void);
 #pragma GCC diagnostic ignored "-Wmissing-declarations"
 #pragma GCC diagnostic ignored "-Wmissing-prototypes"
 
+struct hdr {
+  uint32_t got_sym_start;
+  uint32_t got_start;
+  int got_size;
+  uint32_t data_sym_start;
+  uint32_t data_start;
+  int data_size;
+  uint32_t bss_start;
+  int bss_size;
+  uint32_t reldata_start;
+};
+
+struct reldata {
+  int len;
+  int data[];
+};
+
 __attribute__ ((section(".start"), used))
 __attribute__ ((weak))
 __attribute__ ((noreturn))
-__attribute__ ((naked))
-void _start(void* text_start __attribute__((unused)),
-            void* mem_start __attribute__((unused)),
+void _start(void* text_start,
+            void* mem_start,
             void* memory_len __attribute__((unused)),
             void* app_heap_break __attribute__((unused))) {
-  /*
-   * 1. Setup r9 to point to the GOT (assumes the kernel places the GOT right
-   *    above the stack)
-   * 2. Call `main`
-   * 3. Loop on `yield` forever
-   */
-  asm volatile ("mov r9, sp; \
-                 bl main; \
-                 1: bl yield; \
-                 b 1b");
+
+  // Allocate stack and data. `brk` to STACK_SIZE + got_size + data_size +
+  // bss_size from start of memory
+  uint32_t stacktop = (uint32_t)mem_start + STACK_SIZE;
+  struct hdr* myhdr = (struct hdr*)text_start;
+
+  {
+    uint32_t heap_size = myhdr->got_size + myhdr->data_size + myhdr->bss_size;
+    memop(0, stacktop + heap_size);
```

Comment:
I think you should add a [`memop(10, stacktop);`](https://github.com/helena-project/tock/blob/master/doc/syscalls/memop.md#operation-type-10-debug-specify-stack-location) here. Also probably a [`memop(11, stacktop + heap_size)`](https://github.com/helena-project/tock/blob/master/doc/syscalls/memop.md#operation-type-11-debug-specify-heap-location).

Reply:
Related comment, `heap_size` is a bit of a misleading variable name. I don't think the `got`, `data`, and `bss` sections are traditionally thought of as being on the heap?

(Unfortunately, looking for the "right name" seems to indicate that the .data and .bss are the initialized and uninitialized data segments respectively in the "data" section, which is an awkward aliasing of the word data, maybe `all_data_segments_size`?)